### PR TITLE
[docs] add fsdp_tips.rst

### PR DESCRIPTION
**Pull request by sshleifer**
This is a dumping ground to collect things we want to document about FSDP. 

Please comment with anything random that comes to mind.

Once there have been no changes to FSDP for 48 hrs (or some other proxy for stability), I will finalize this and set it to Ready For Review.

### CODE REVIEW

File: fairscale/nn/data_parallel/README_fsdp.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+```
+will use an incorrect norm (the norm over all params in a shard) when clipping gradients.
+To overcome this, you can either call
+`sharded_module.clip_grad_norm(1.0)`
+which does the extra computation required to compute the norm properly, or use `torch.nn.utils.clip_grad_value_`.
+```
+
+
+
+#### Misc
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/README_fsdp.md:77]: should this include the new wrap/auto_wrap feature?

Comment (sshleifer) [reply fairscale/nn/data_parallel/README_fsdp.md:77]: Yes leave desired `md` in comment else I can take a pass tomorrow.

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/README_fsdp.md:77]:
```
#### The `enable_wrap` context

There are two cases where the `enable_wrap` context can be useful:

* When you'd like to apply the same parameters to all child modules that you wrap with FSDP. Calling the `wrap` function within the said context will save you from passing the same set of FSDP parameters explicitly.
* When wrapping large models that does NOT fit within the CPU memory. I.e. you don't first create the full model and then traverse it to wrap it with FSDP at different parts. Instead, you create a wrapped instance of the model incrementally as you build up the model, allowing large modules to be sharded in-place.

example:
        with enable_wrap(**fsdp_params):
            # Wraps layer in FSDP by default if within context
            self.l1 = wrap(torch.nn.Linear(5, 5))
            # Wraps children modules by default based on min_num_params
            self.l2 = auto_wrap(TransformerBlock(), min_num_params=1e8)
```

**Comment by min-xu-ai** [comment]
One more thing to be added to the doc:

```
If model weight initialization happens deterministically after sharding, the final weights will be comprised of N piece of identical shards. This would have negative effects on the model initial state.
```

@myleott @sshleifer Does the above looks good? If so, I can also add it to the doctoring of FSDP. We generate doc from the doctoring.

**Comment by sshleifer** [comment]
These are not at all perfect, please feel free to push changes to this branch or comment.
I made a separate `fsdp_tips.rst` and cross-linked with `fsdp.rst` to avoid one crowding out the other on the same page. There may be a nicer way to do this in `rst`, (e.g. linking to section 2 at the top of 1 document).

**Comment by sshleifer** [comment]
[fsdp_tips.html.zip](https://github.com/facebookresearch/fairscale/files/6091635/fsdp_tips.html.zip)

File: docs/source/api/nn/fsdp_tips.rst [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+* FSDP enables training 13B parameter models on 8 GPUs and 175B parameter models on 128 GPUs. When using the ``cpu_offload=True`` option, it's possible to train 1T parameter models on 256 GPUs.
+
+
+General usage notes
+--------------------
+
+-  For best memory efficiency use ``auto_wrap`` to wrap each layer in your network with ``FSDP`` and set ``reshard_after_forward=True``
+-  For best training speed set ``reshard_after_forward=False`` (wrapping each layer is not required, but will improve speed further)
+-  If you're using ``torch.cuda.amp.autocast`` for mixed precision, that's fully compatible with the FSDP wrapper, just set ``mixed_precision=True``
+-  Uf combining with `activation checkpointing <https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/misc/checkpoint_activations.py>`__,
```
Comment (Vittorio-Caggiano) [inline docs/source/api/nn/fsdp_tips.rst:34]: typo: `if`

File: docs/source/api/nn/fsdp_tips.rst [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+The key insight to unlock full parameter sharding is that we can decompose the
+`all-reduce <https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#allreduce>`__
+operation in DDP into separate
+`all-gather <https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#allgather>`__
+and
+`reduce-scatter <https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#reducescatter>`__
+operations:
+
+.. |Figure 1| image:: https://user-images.githubusercontent.com/231798/108780259-26870a00-7536-11eb-890d-51720f39d098.png
```
Comment (Vittorio-Caggiano) [inline docs/source/api/nn/fsdp_tips.rst:56]: we have a higher quality image with the same font/format as the next one, do you want to use that? 
image

**Review by Vittorio-Caggiano — COMMENTED** [review]
just a small suggestion. 
Would it possible to add also a small tutorial?

**Comment by myleott** [comment]
> Would it possible to add also a small tutorial?

Yes, this is a great idea 😄

Just to cross-reference and not lose track, Min also suggested the tutorial should include custom weight init inside a `summon_full_params` context: https://github.com/facebookresearch/fairscale/issues/454#issuecomment-792223593

File: docs/source/api/nn/fsdp_tips.rst [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+    import torch
+    from fairscale.nn.wrap import auto_wrap, enable_wrap, wrap
+    from fairscale.nn.data_parallel import FullyShardedDataParallel
+    from fairscale.utils.testing import DummyProcessGroup
+    tfmr = torch.nn.Transformer(num_encoder_layers=2, num_decoder_layers=2)
+
+    group = DummyProcessGroup(rank=0, size=1)
+    fsdp_params = dict(mixed_precision=True, flatten_parameters=True)
+    with enable_wrap(process_group=group, **fsdp_params):
```
Comment (myleott) [inline docs/source/api/nn/fsdp_tips.rst:122]: switch to new syntax:
```python
with enable_wrap(wrapper_cls=FSDP, ...)
```

File: docs/source/api/nn/fsdp_tips.rst [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+-  Sharded parameters are downcast to ``fp16`` before ``forward``, promoted to ``fp32`` after forward.
+-  buffers are kept in ``fp16``, unless ``buffer_dtype=torch.float32`` is passed. Buffers are not sharded regardless of arguments.
+-  By default, gradients will be computed and reduced
+-  ``fp32_reduce_scatter=True`` controls the quantization of the gradient communication
+-  If ``torch.amp.autocast`` is enabled it will override the output dtypes of some operations, like ``BatchNorm2D``
+
+
+Auto-wrap
+~~~~~~~~~
+Auto wrapping sub-modules with ``FSDP`` is a convenient way to improve training speed by overlapping the all-gather step across the forward passes of different submodules.
```
Comment (myleott) [inline docs/source/api/nn/fsdp_tips.rst:108]: Add:
```
It also improves memory efficiency by freeing gathered parameters after each layer finishes executing.
```

File: docs/source/api/nn/fsdp_tips.rst [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+
+Mixed Precision
+---------------
+
+When ``mixed_precision=True``:
+
+-  Sharded parameters are downcast to ``fp16`` before ``forward``, promoted to ``fp32`` after forward.
+-  buffers are kept in ``fp16``, unless ``buffer_dtype=torch.float32`` is passed. Buffers are not sharded regardless of arguments.
+-  By default, gradients will be computed and reduced
```
Comment (myleott) [inline docs/source/api/nn/fsdp_tips.rst:101]:
```
By default, gradients will be computed and reduced in fp16. If FP32 reductions are important, set ``fp32_reduce_scatter=True``
```

File: docs/source/api/nn/fsdp_tips.rst [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            backward pass for layer_i
+            discard full weights for layer_i
+            reduce-scatter gradients for layer_i
+
+Saving and Loading
+------------------
+
+There are two ways to load and save FSDP instances,
+
+- ``state_dict()`` returns a dictionary containing all parameters, which can be loaded with ``load_local_state_dict()``
```
Comment (myleott) [inline docs/source/api/nn/fsdp_tips.rst:90]: nit: this should be `load_state_dict` at the end

File: docs/source/api/nn/fsdp_tips.rst [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+Using CPU RAM
+-------------
+
+``move_grads_to_cpu`` and ``cpu_offload`` control which tensors get
+moved to CPU.
+
+-  ``cpu_offload`` moves weights to CPU when they are not being used.
+-  ``move_grads_to_cpu`` moves gradients to CPU. The use of this option
+   requires that the optimizer has a copy of the model parameters on
+   CPU.
```
Comment (myleott) [inline docs/source/api/nn/fsdp_tips.rst:155]: How about:
```
``move_grads_to_cpu`` moves gradients to CPU so that the optimizer step also happens on CPU. This option requires ``cpu_offload=True``.
```

**Review by myleott — COMMENTED** [review]
This looks great, thanks @sshleifer! I made some comments below, but I think we can ship this and iterate. What do you think @min-xu-ai?

**Comment by min-xu-ai** [comment]
> I think we can ship this and iterate. What do you think @min-xu-ai?

Yeah, totally.
